src/celestia/celx_misc.cpp

```diff
@@ -261,132 +241,109 @@ static int image_new(lua_State* l, Image* i)
 }
 #endif
 
-static Image* to_image(lua_State* l, int index)
-{
-    Image** image = static_cast<Image**>(lua_touserdata(l, index));
-
-    return image != nullptr ? *image : nullptr;
-}
+#define this_image getThis<Image*>()
 
-static Image* this_image(lua_State* l)
+int celxClassId(Image*)
 {
-    Image* image = to_image(l,1);
-    if (image == nullptr)
-        Celx_DoError(l, "Bad image object!");
-
-    return image;
+    return Celx_Image;
 }
 
 static int image_getheight(lua_State* l)
 {
-    Celx_CheckArgs(l, 1, 1, "No arguments expected for image:getheight()");
+    CelxLua celx(l);
 
-    Image* image = this_image(l);
-    lua_pushnumber(l, image->getHeight());
-    return 1;
+    celx.checkArgs(1, 1, "No arguments expected for image:getheight()");
+
+    auto image = *celx.this_image;
+    return celx.push(image->getHeight());
 }
 
 static int image_getwidth(lua_State* l)
 {
-    Celx_CheckArgs(l, 1, 1, "No arguments expected for image:getwidth()");
+    CelxLua celx(l);
 
-    Image* image = this_image(l);
-    lua_pushnumber(l, image->getWidth());
-    return 1;
+    celx.checkArgs(1, 1, "No arguments expected for image:getwidth()");
+
+    auto* image = *celx.this_image;
+    return celx.push(image->getWidth());
 }
 
 static int image_tostring(lua_State* l)
 {
-    // TODO: print out the actual information about the image
-    lua_pushstring(l, "[Image]");
-
-    return 1;
+    CelxLua celx(l);
+    auto image = *celx.this_image;
+    std::string s = fmt::sprintf("[Image:%dx%d]", image->getWidth(), image->getHeight());
+    return celx.push(s.data());
 }
 
 void CreateImageMetaTable(lua_State* l)
 {
-    Celx_CreateClassMetatable(l, Celx_Image);
-
-    Celx_RegisterMethod(l, "__tostring", image_tostring);
-    Celx_RegisterMethod(l, "getheight", image_getheight);
-    Celx_RegisterMethod(l, "getwidth", image_getwidth);
-
-    lua_pop(l, 1); // remove metatable from stack
-}
-
-// ==================== Texture ============================================
-
-int texture_new(lua_State* l, Texture* t)
-{
-    Texture** ud = static_cast<Texture**>(lua_newuserdata(l, sizeof(Texture*)));
-    *ud = t;
-
-    Celx_SetClass(l, Celx_Texture);
+    CelxLua celx(l);
 
-    return 1;
-}
+    celx.createClassMetatable(Celx_Image);
 
-static Texture* to_texture(lua_State* l, int index)
-{
-    Texture** texture = static_cast<Texture**>(lua_touserdata(l, index));
+    celx.registerMethod("__tostring", image_tostring);
+    celx.registerMethod("getheight", image_getheight);
+    celx.registerMethod("getwidth", image_getwidth);
 
-    return texture != nullptr ? *texture : nullptr;
+    celx.pop(1); // remove metatable from stack
 }
 
-static Texture* this_texture(lua_State* l)
-{
-    Texture* texture = to_texture(l,1);
-    if (texture == nullptr)
-        Celx_DoError(l, "Bad texture object!");
+// ==================== Texture ============================================
 
-    return texture;
-}
+#define this_tex getThis<Texture*>()
```

Comment:
J U S T  N O  W A Y.

Preprocessor must not be used this way.

Reply:
Yes, I knoe, wrong style. But what is the best way of preprocessor use?

Reply:
in C++ it's better to avoid preprocessor. and definitely it shouldn't be used just to save some key strokes.

Reply:
Qt programmers certainly didn't use this rule... :-)